<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Проверка-на-тестовой-выборке" data-toc-modified-id="Проверка-на-тестовой-выборке-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Проверка на тестовой выборке</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

В ходе выполнения данной работы будем придерживаться следующийх этапов:

1) загрузим данные и проверим их корректоность. Если будут обнаружены какие-то ошибки или аномалии, то мы их исправим.

2) проведем теоретическое доказательство ответов на поставленные ниже вопросы

3) разработаем алгоритм преобразования

4) реализуем данный алгоритм и проверим его на тестовой выборке

5) дадим общий вывод.

## Загрузка данных

In [1]:
# импортируем необходимые библиотеки
import pandas as pd
import numpy as np

from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

In [2]:
# прочитаем данные
data = pd.read_csv('....csv')

In [3]:
# выведем первые 10 строк на экран
data.head(10)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0
5    1     43.0   41000.0            2                  1
6    1     39.0   39700.0            2                  0
7    1     25.0   38600.0            4                  0
8    1     36.0   49700.0            1                  0
9    1     32.0   51700.0            1                  0

In [4]:
# определим размер таблицы
data.shape

(5000, 5)

In [5]:
# выведем общую информацию
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


В ходе анализа таблицы получили следующее:

1) в таблице 5 столбцов и 5000 строк

2) в таблице следующие столбцы:

    - пол: тип данных int
    - возраст: тип данных float
    - зарплата: тип данных float
    - члены семьи: тип данных int
    - страховые выплаты(их количество): тип данных int

In [6]:
# выведем названия колонок
data.columns

Index(['Пол', 'Возраст', 'Зарплата', 'Члены семьи', 'Страховые выплаты'], dtype='object')

In [7]:
# переименуем колонки в соответствии с общепринятыми требованиями
data = data.rename(columns={'Пол': 'gender', 'Возраст': 'age', 'Зарплата': 'salary', 'Члены семьи': 'family_members', 'Страховые выплаты': 'insurance_payments'})

In [8]:
# определим количество пропусков
data.isna().sum()

gender                0
age                   0
salary                0
family_members        0
insurance_payments    0
dtype: int64

In [9]:
#  посмотрим на столбец age
data['age'].value_counts()

19.0    223
25.0    214
31.0    212
26.0    211
27.0    209
22.0    209
32.0    206
28.0    204
29.0    203
30.0    202
23.0    202
21.0    200
20.0    195
36.0    193
33.0    191
24.0    182
35.0    179
34.0    177
37.0    147
39.0    141
38.0    139
41.0    129
18.0    117
40.0    114
42.0     93
43.0     77
44.0     74
45.0     73
46.0     60
47.0     47
49.0     37
50.0     27
48.0     26
52.0     22
51.0     21
53.0     11
55.0      9
54.0      7
56.0      5
59.0      3
60.0      2
58.0      2
57.0      2
65.0      1
61.0      1
62.0      1
Name: age, dtype: int64

In [10]:
#  посмотрим на столбец gender
data['gender'].value_counts()

0    2505
1    2495
Name: gender, dtype: int64

In [11]:
#  посмотрим на столбец salary
data['salary'].describe()

count     5000.000000
mean     39916.360000
std       9900.083569
min       5300.000000
25%      33300.000000
50%      40200.000000
75%      46600.000000
max      79000.000000
Name: salary, dtype: float64

In [12]:
#  посмотрим на столбец family_members
data['family_members'].value_counts()

1    1814
0    1513
2    1071
3     439
4     124
5      32
6       7
Name: family_members, dtype: int64

In [13]:
#  посмотрим на столбец insurance_payments
data['insurance_payments'].value_counts()

0    4436
1     423
2     115
3      18
4       7
5       1
Name: insurance_payments, dtype: int64

In [14]:
# изменим тип данных в поле Возраст на int64
data['age'] = data['age'].astype('int64')
data['salary'] = data['salary'].astype('int64')

In [15]:
# проверим дубликаты
data.duplicated().sum()

153

В наших данных 153 дубликата. Удалим их.

In [16]:
# удаление дубликатов
data.drop_duplicates(inplace=True)

In [17]:
# проверим удаление дубликатов
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4847 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype
---  ------              --------------  -----
 0   gender              4847 non-null   int64
 1   age                 4847 non-null   int64
 2   salary              4847 non-null   int64
 3   family_members      4847 non-null   int64
 4   insurance_payments  4847 non-null   int64
dtypes: int64(5)
memory usage: 227.2 KB


### Вывод

В ходе выполнения данного раздела, мы прочитали данные из файла, проверили их и получили следующее:

1) переименовали названия столбцов для удобства работы в дальнейшем

2) определели размер таблицы (5 столбцов и 4847 строк, после удаления дубликатов), посмотрели состав колонок (gender, age, salary, family_members, insurance_payments), ошибок в данных обнаружено не было

3) отсутствую пропуски

4) изменили тип данных в столбце возраст, так как в данном столбце возраст хранится целым числом.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** при умножении признаков на обратимую матрицу качество линейной регрессии не меняется

**Обоснование:** обоснуем данное утверждение формулой расчета весов для линейной регресси.
Для того, чтобы правильно решить данную формулу уточним основные свойства матриц:

1) $(AB)^{-1} = (B)^{-1}(A)^{-1}$

2) $(AB)^T = B^T A^T$

3) $AA^{-1} = E$

Предсказания:

$$
a = Xw
$$

Вектор весов линейной регрессии:

$$
w = (X^T X)^{-1}X^T y
$$

Представим новую матрицу признаков X1 как произведение исходной матрицы пррзнаков Х на обратимую матрицу Р:

$$
X1 = XP
$$

Далее подставим полученное значение Х1 в вектор весов линейной регрессии w1:

$$
w1 = ((XP)^T(XP))^{-1}(XP)^Ty
$$

Будем раскрывать скобки постепенно в соответствии с указанными выше свойствами матриц. Для начала раскрем первую скобку $(XP)^Y$:

$$
w1 = (P^T X^T XP)^{-1}(XP)^Ty
$$

Для наглядности перегруппируем множители в скобке:

$$
w1 = (P^T(X^T X)P)^{-1}(XP)^Ty
$$

В соответствии со свойставми матриц раскроем скобки $(P^T(X^TX)P)^{-1}$:

$$
w1 = P^{-1}(X^TX)^{-1}(P^T)^{-1}(XP)^Ty
$$

Раскроем последнюю скобку $(XP)^T$:

$$
w1 = P^{-1}(X^TX)^{-1}(P^T)^{-1}P^TX^Ty
$$

Так как по условию задачи матрица Р обратимая, то исходя из свойства матриц номер 3, указанных выше. То есть произведение матрицы на обратимую матрицу должно быть равно единичной матрице Е. Получаем, что $(P^T)^{-1}P^T = E$. Преобразуем наше выражение:

$$
w1 = P^{-1}(X^TX)^{-1}EX^Ty = P^{-1}(X^TX)^{-1}X^Ty
$$

Как можно удидеть, в данной формуле мы получили формулу для вектора весов линейной регрессии $w = (X^T X)^{-1}X^T y$, таком образом получаем:

$$
P^{-1}(X^TX)^{-1}X^Ty = P^{-1}w
$$

Если в формулу предскания а подставить значения X1 и w1, то мы получим следующее предсказание а1 для новой модели:

$$
a1 = X1w1 = XPP^{-1}w
$$

Так как Р по условию задачи обратима, то исхлдя из своийств матриц мы можем заменить выржение $PP^{-1}$ на единичную матрицу Е, тогда:

$$
a1 = XPP^{-1} = XEw = Xw = a
$$

Таким образом мы доказали, что предсказания а1 для матрицы признаков, умноженное на обратимую матрицу Р, равны предсказаниям а. Исходя из этого можно сделать вывод, что при умножении признаков на обратимую матрицу качество модели линейной регрессии не изменится. Таким образом, мы можем разработать алгоритм для шифрования персональных данных клиентов.

## Алгоритм преобразования

**Алгоритм**

В качестве алгоритма предобразования можно предложить следующее: 

1) разделим данные для обучения

2) создадим случайную матрицу, проверим ее на обратимость

3) обучим модель на исходной матрице и рассчитаем значение r2_score

4) случайную матрицу умножим на признаки исходной матрицы

5) обучим модель на полученных признаках и рассчитаем значение r2_score

6) сравним полученные метрики между собой.

Если характеристики r2_score будут одинаковыми, то данный алгоритм можно использовать для шифрования данных клиентов.

**Обоснование**

Исходя из теоретического доказательства формулы выше, можно предположить, что благодаря предложенному аалгоритму преобразования мы должны получить похожие значения метрик. Для подтверждения нашей гипотезы будет сравнивать между собой r2_score исходной матрицы и r2_score случайной матрицы (зашифрованные признаки). Если значения не будут различаться или различаться незначительно, то мы подтвердим доказательство формулы выше. Это будет означать, что разработанный нами алгоритм можно применять для шифровани данных.

## Проверка алгоритма

In [18]:
# разделим данные
features = data.drop('insurance_payments', axis=1)
target = data['insurance_payments']

In [19]:
# разделим полученную выборку на обучающую и тестовую
features_train, features_test, target_train, target_test = train_test_split(features, target, test_size=0.25, random_state=12345)

In [20]:
# проверим, что размеры признаков и целевых признаков в выборках совпадают
print(f'Обучающая выборка: {features_train.shape}, {target_train.shape}')
print(f'Тестовая выборка: {features_test.shape}, {target_test.shape}')

Обучающая выборка: (3635, 4), (3635,)
Тестовая выборка: (1212, 4), (1212,)


In [21]:
# создадим рандомную матрицу
matrix = np.random.normal(size = (features.shape[1], features.shape[1]))

In [22]:
# проверим ее на обратимость
matrix = np.linalg.inv(matrix)

In [23]:
# обучим модель на исходных, незашифрованных признаках
model = LinearRegression()
model.fit(features_train, target_train)
predictions_train = model.predict(features_train)
r2_score_train = r2_score(target_train, predictions_train)
print(f'r2_score на исходных, незашифрованных признаках: {r2_score_train}')

r2_score на исходных, незашифрованных признаках: 0.43215820265809746


In [24]:
# умножим исходную матрицу на случайную 
transform_features_train = features_train @ matrix

In [25]:
# обучим модель на новых, зашифрованных признаках
model.fit(transform_features_train, target_train)
predictions_train_transform = model.predict(transform_features_train)
r2_score_train_transform = r2_score(target_train, predictions_train_transform)
print(f'r2_score на новых, зашифрованных признаках: {r2_score_train_transform}')

r2_score на новых, зашифрованных признаках: 0.43215820265810523


In [26]:
# определим, есть ли разность между полученными данными
print('Разность между признаками', r2_score_train - r2_score_train_transform)

Разность между признаками -7.771561172376096e-15


**Вывод:** как видим, разность между признаками исходной и матрицы, полученной умножением на случайную матрицу (обратимую матрицу) практически отсутсвует. Это говорит о том, что на практике мы подтвердили теоретическое доказательсво того, что при умножении признаков на обратимую матрицу качество линейной регрессии не меняется.

###  Проверка на тестовой выборке

Проверим наши модели на тестовой выборке.

In [27]:
model = LinearRegression()
model.fit(features_test, target_test)
predictions_test = model.predict(features_test)
r2_score_test = r2_score(target_test, predictions_test)
print(f'r2_score на исходных, незашифрованных признаках: {r2_score_test}')

r2_score на исходных, незашифрованных признаках: 0.425552801076332


In [28]:
# умножим исходную матрицу на случайную 
transform_features_test = features_test @ matrix

In [29]:
# обучим модель на новых, зашифрованных признаках
model.fit(transform_features_test, target_test)
predictions_test_transform = model.predict(transform_features_test)
r2_score_test_transform = r2_score(target_test, predictions_test_transform)
print(f'r2_score на новых, зашифрованных признаках: {r2_score_test_transform}')

r2_score на новых, зашифрованных признаках: 0.4255528010763243


In [30]:
# определим, есть ли разность между полученными данными
print('Разность между признаками', r2_score_test - r2_score_test_transform)

Разность между признаками 7.66053886991358e-15


Как видим и на тестовой выборке мы получили похожие знаачения r2_score и похожую разницу между показателями. Таким образом, мы еще раз подтвердили, что что при умножении признаков на обратимую матрицу качество линейной регрессии не меняется. Наш алгоритм можно использовать для шифрования данных клиентов

## Вывод

В ходе выполнения работы, мы:

1) проанализировали данные, увидели, что данные впорядке, за исключение дубликатов, которые были устранены

2) выдвинули предположение о том, что при умножении признаков на обратимую матрицу качество линейной регрессии не меняется

3) теоретически доказали правильность данного утверждения

4) разработали алгоритм, на основании которого можно практически доказать верность утверждения

5) подготовили данные для обученяи моделей

6) обучили модели и проверили поведение модели, при умножении исходных признаком на случайную обратимую матрицу

7) получили одинаковые данные и таким образом на практике доказали, что при умножении признаков на обратимую матрицу качество линейной регрессии не меняется

8) на тестовой выборке проверили верность наших расчетов и доказательств

Такми образом, при умножении признаков на обратимую матрицу качество линейной регрессии не меняется и разработанный нами алгоритм можно применять для шифровани данных клиентов